src/replay/scheduler.zig

```diff
@@ -0,0 +1,294 @@
+
+        return .{
+            .allocator = allocator,
+            .batches = batches,
+            .thread_pool = pool,
+            .results = channel,
+            .locks = .{},
+            .batches_started = 0,
+            .batches_finished = 0,
+            .exit = std.atomic.Value(bool).init(false),
+            .failure = null,
+        };
+    }
+
+    pub fn deinit(self: TransactionScheduler) void {
+        var batches = self.batches;
+        for (batches.items) |batch| batch.deinit(self.allocator);
+        batches.deinit(self.allocator);
+
+        var channel = self.results;
+        channel.deinit();
+
+        self.thread_pool.deinit(self.allocator);
+        self.locks.deinit(self.allocator);
+    }
+
+    pub fn addBatch(self: *TransactionScheduler, batch: ResolvedBatch) Allocator.Error!void {
+        try self.batches.append(self.allocator, batch);
+    }
+
+    pub fn addBatchAssumeCapacity(self: *TransactionScheduler, batch: ResolvedBatch) void {
+        self.batches.appendAssumeCapacity(batch);
+    }
+
+    pub fn poll(self: *TransactionScheduler) !ConfirmSlotStatus {
+        // collect results
+        while (self.results.tryReceive()) |maybe_err| {
+            self.batches_finished += 1;
+            if (maybe_err) |err| {
+                self.exit.store(true, .monotonic);
+                self.failure = .{ .invalid_transaction = err };
+            }
+        }
+
+        // process results
+        switch (self.thread_pool.pollFallible()) {
+            .done => {
+                assert(self.batches_started == self.batches_finished);
+                if (self.failure) |f| {
+                    return .{ .err = f };
+                } else if (self.batches.items.len != self.batches_started) {
+                    if (try self.tryScheduleSome()) |err| {
+                        self.exit.store(true, .monotonic);
+                        self.failure = .{ .invalid_transaction = err };
+                    }
+                    return .pending;
+                } else {
+                    assert(self.batches.items.len == self.batches_finished);
+                    return .done;
+                }
+            },
+            .pending => return .pending,
+            .err => |err| return err,
+        }
+    }
+
+    fn tryScheduleSome(self: *TransactionScheduler) !?TransactionError {
+        while (self.batches.items.len > self.batches_started) {
+            const batch = self.batches.items[self.batches_started];
+            self.locks.lockStrict(self.allocator, batch.accounts) catch |e| {
+                switch (e) {
+                    error.LockFailed => if (self.batches_started == self.batches_finished) {
+                        return .AccountInUse;
+                    } else {
+                        break;
+                    },
+                    else => return e,
+                }
+            };
+            assert(try self.thread_pool.trySchedule(self.allocator, .{
```

Comment:
Can we get a comment on why this assertion should hold true? It's not obvious to me.

Reply:
86283f5aff1b1433a689e3b822f45d2fbea3f5ec